cloudwatch_metrics_collector/src/metrics_collector.cpp

```diff
@@ -101,26 +135,98 @@ int MetricsCollector::RecordMetrics(
+    Aws::CloudWatchMetrics::Utils::MetricObject metric_object{metric_msg->metric_name,
                                                               metric_msg->value,
                                                               metric_msg->unit,
-                                                              GetMetricDataEpochMillis(*metric_msg),
+                                                              GetMetricDataEpochMillis(metric_msg->time_stamp),
                                                               dimensions,
                                                               this->storage_resolution_.load()};
     bool batched = metric_service_->batchData(metric_object);
 
     if (!batched) {
-      AWS_LOGSTREAM_ERROR(__func__, "Failed to record metric");
+      AWS_LOGSTREAM_ERROR(__func__, "Failed to record ROS monitoring message");
+    } else {
+      ++batched_count;
+    }
+  }
+  return batched_count;
+}
+
+int MetricsCollector::RecordMetrics(
+  metrics_statistics_msgs::msg::MetricsMessage::UniquePtr msg)
+{
+  int batched_count = 0;
+
+  AWS_LOGSTREAM_DEBUG(__func__, "Received metric with " << msg->statistics.size() << " types of statistics");
+  AWS_LOGSTREAM_DEBUG(__func__, "Recording metric with name=[" << msg->measurement_source_name << "]");
+
+  std::map<StatisticValuesType, double> statistic_values;
+  double stddev;
+  for (const auto & statistic : msg->statistics) {
+    if (statistic.data_type == StatisticDataType::STATISTICS_DATA_TYPE_AVERAGE) {
+      statistic_values[StatisticValuesType::SUM] = statistic.data;
+    } else if (statistic.data_type == StatisticDataType::STATISTICS_DATA_TYPE_MAXIMUM) {
+      statistic_values[StatisticValuesType::MAXIMUM] = statistic.data;
+    } else if (statistic.data_type == StatisticDataType::STATISTICS_DATA_TYPE_MINIMUM) {
+      statistic_values[StatisticValuesType::MINIMUM] = statistic.data;
+    } else if (statistic.data_type == StatisticDataType::STATISTICS_DATA_TYPE_STDDEV) {
+      stddev = statistic.data;
+    } else if (statistic.data_type == StatisticDataType::STATISTICS_DATA_TYPE_SAMPLE_COUNT) {
+      statistic_values[StatisticValuesType::SAMPLE_COUNT] = statistic.data;
     }
+  }
+  try {
+    double & sum = statistic_values.at(StatisticValuesType::SUM);
+    sum *= statistic_values.at(StatisticValuesType::SAMPLE_COUNT);
+  } catch (const std::out_of_range & exception) {
+    statistic_values.erase(StatisticValuesType::SUM);
+  }
 
-    batched_count++;
+  std::map<std::string, std::string> dimensions;
+  dimensions["metric_type"] = msg->metrics_source;
+  dimensions["metric_unit"] = msg->unit;
+
+  // create a MetricObject with message parameters to batch
+  MetricObject metric_object(
+    msg->measurement_source_name,
+    "msec",
+    GetMetricDataEpochMillis(msg->window_start),
+    GetMetricDataEpochMillis(msg->window_stop) - GetMetricDataEpochMillis(msg->window_start),
```

Comment:
Is this no longer needed given the duration findings?

Reply:
Right. I have removed them now.